### ISSUE TITLE
Internal error after refining a tactic @ MetaVars.hs:267

### ISSUE THREAD

**Issue by m0davis** [agda/agda#2368]
This one seemed a bit too daunting for me to simplify. It uses agda-prelude.
```agda
open import Tactic.Nat
open import Prelude.Nat
open import Prelude.Empty
open import Prelude.Bool
open import Prelude.Equality

lemma : ∀ {m n : Nat} → ¬ IsTrue (lessNat m n) → ¬ IsTrue (lessNat n m) → m ≡ n
lemma {zero} {zero} m≮n n≮m = refl
lemma {zero} {suc n} m≮n n≮m = ⊥-elim (m≮n true)
lemma {suc m} {zero} m≮n n≮m = ⊥-elim (n≮m true)
lemma {suc m} {suc n} m≮n n≮m = by (lemma {m} {n} {!m≮n!} {!n≮m!})
```
Refine (C-c C-r) the hole containing `m≮n`, then refine (without reloading) the last hole. The result is:
```
An internal error has occurred. Please report this as a bug.
Location of the error: src/full/Agda/TypeChecking/Monad/MetaVars.hs:267
```

**Comment by andreasabel** [comment]
> It uses agda-prelude.

That's for @UlfNorell then, I suppose.

**Comment by m0davis** [comment]
> It uses agda-prelude.
> > That's for @UlfNorell then, I suppose.

Sorry. I'll buckle down and make a simplified error case.

**Comment by m0davis** [comment]
Heh. It wasn't nearly so hard as I imagined it would be:
```agda
open import Agda.Builtin.Reflection

macro
  mac : Term → Term → TC _
  mac = {!!}

foo : Set → Set → Set
foo = mac (foo {!Term!} {!Term!})
```
Load it, refine the first `Term`, then (without reloading, of course), try to refine the second `Term`, and voilà, an error at MetaVars.hs:267.

**Comment by andreasabel** [comment]
Thanks.
I tried to simulate a similar interaction pattern without reflection:
```agda
data C : Set where
  c : C → C → C
  d : C

data D : Set where
  c : D

foo = c {!d!} {!d!}
```
There is a postponed type checking problem `_9 := c ? ? :? _8`.  
If I try to give a `d`, I get
```
No type nor action available for hole ?0. Possible cause: the hole
has not been reached during type checking (do you see yellow?)
```
This is because the type checker never reached the two interaction metas; they are unnumbered question marks in the postponed type checking problem constraint.

In your example, Agda did reach the interaction points during type checking, but still there is a postponed type checking problem (`_9`):
```
?0 : Term → Term → TC .Agda.Builtin.Unit.⊤
?1 : Set
?2 : Set
_6 : Set → Set → Set  [ at /home/abel/tmp/agda-issue-2327/test/bugs/Issue2368-orig.agda:10,7-33 ]
_10 : Term  [ at /home/abel/tmp/agda-issue-2327/test/bugs/Issue2368-orig.agda:10,12-33 ]
_12 : Set → Set → Set  [ at /home/abel/tmp/agda-issue-2327/test/bugs/Issue2368-orig.agda:7,9-13 ]
Failed to solve the following constraints:
  _11 := unquote (mac _10)
  _9 := quoteTerm (foo ? ?) :? Term
```
This problem has two interaction points, which connect to metas ?1 and ?2, and it is blocked on ?1, ?2.  As soon both of these are instantiated, it resumes the postponed type checking problem, encounters the interaction points, tries to allocate fresh metas to connect with them.  However, these interaction points have been removed from the store, since they have been solved.  Agda panics. 
```haskell
          | A.QuoteTerm _ <- unScope q ->
             do (et, _)   <- inferExpr (namedThing e)
                et'       <- etaContract =<< instantiateFull et
                let metas = allMetas et'
                case metas of
                  _:_ -> postponeTypeCheckingProblem (CheckExpr e0 t) $ andM $ map isInstantiatedMeta metas
                  []  -> do
                    q  <- quoteTerm et'
                    ty <- el primAgdaTerm
                    coerce q ty t
```
The interaction point machinery is not prepared for the following scenario:
1. A term is successfully type-checked and connects ips to metas.
2. The ips are solved.
3. The type checker revisits the term and tries again to connect the ips.

The code for `quoteTerm` gullibly relies on such a scenario.

**Comment by andreasabel** [comment]
A solution would be to never remove interaction points from the store, rather just mark them as solved.
I am implementing it.
